rules/TypeDeclaration/TypeNormalizer.php

```diff
@@ -52,16 +52,17 @@ public function normalizeArrayOfUnionToUnionArray(Type $type, int $arrayNesting
             $this->collectedNestedArrayTypes = [];
         }
 
-        if ($type->getItemType() instanceof ArrayType) {
+        $itemType = $type->getIterableValueType();
+        if ($itemType instanceof ArrayType) {
             ++$arrayNesting;
-            $this->normalizeArrayOfUnionToUnionArray($type->getItemType(), $arrayNesting);
-        } elseif ($type->getItemType() instanceof UnionType) {
-            $this->collectNestedArrayTypeFromUnionType($type->getItemType(), $arrayNesting);
+            $this->normalizeArrayOfUnionToUnionArray($itemType, $arrayNesting);
+        } elseif ($itemType instanceof UnionType) {
```

Comment:
this can be `IntersectionType`, see PR patch that happen on downgrade:

- https://github.com/rectorphp/rector-downgrade-php/pull/237

due to downgrade error:

- https://github.com/rectorphp/rector-src/actions/runs/11936299874/job/33269627769#step:10:21

until we have more fixture tests, I think we can keep using strict `ArrayType` and `ConstantArrayType` to avoid regression.

Reply:
ok, lets do it less aggressive: https://github.com/rectorphp/rector-src/pull/6479